File: app/src/main/java/com/hwonchul/movie/presentation/view/MainActivity.java
package com.hwonchul.movie.presentation.view;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;

import com.google.android.material.snackbar.Snackbar;
import com.hwonchul.movie.R;
import com.hwonchul.movie.databinding.ActivityMainBinding;
import com.hwonchul.movie.util.NetworkStatusHelper;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class MainActivity extends AppCompatActivity {
    private ActivityMainBinding binding;

    @Inject
    NetworkStatusHelper helper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setNetworkStatusHelper();
        setStatusBarAndNavigationBar();
    }

    private void setNetworkStatusHelper() {
        // 네트워크 상태 감지기
        helper.observe(this, networkStatus -> {
            // 인터넷 연결됨
            if (networkStatus == NetworkStatusHelper.NetworkStatus.AVAILABLE) {
                final int greenColor =
                        ContextCompat.getColor(getApplication(), android.R.color.holo_green_dark);
                binding.tvNetworkState.setBackgroundColor(greenColor);
                binding.tvNetworkState.setText(R.string.all_network_connected);
                new Handler(Looper.getMainLooper()).postDelayed(() ->
                        binding.tvNetworkState.setVisibility(View.GONE), 2000);
            }
            // 인터넷 연결 끊김
            else if (networkStatus == NetworkStatusHelper.NetworkStatus.UNAVAILABLE) {
                final int redColor =
                        ContextCompat.getColor(getApplication(), android.R.color.holo_red_dark);
                binding.tvNetworkState.setBackgroundColor(redColor);
                binding.tvNetworkState.setText(R.string.all_network_disconnected);
                binding.tvNetworkState.setVisibility(View.VISIBLE);
            }
        });
    }

    private void setStatusBarAndNavigationBar() {
        NavHostFragment navHostFragment =
                (NavHostFragment) getSupportFragmentManager().findFragmentById(R.id.nav_host_fragment);
        NavController navController = navHostFragment.getNavController();
        navController.addOnDestinationChangedListener((__, navDestination, ___) -> {
            final int whiteColor = ContextCompat.getColor(getApplication(), R.color.WHITE);
            final int blackColor = ContextCompat.getColor(getApplication(), R.color.BLACK);
            final int blue700Color = ContextCompat.getColor(getApplication(), R.color.colorPrimary);

            int uiVisibility = 0;

            // 사진보기 화면
            if (navDestination.getId() == R.id.poster) {
                getWindow().setStatusBarColor(blackColor);
                getWindow().setNavigationBarColor(blackColor);
            }
            // 로그인 화면
            else if (navDestination.getId() == R.id.login
                    || navDestination.getId() == R.id.phone_auth
                    || navDestination.getId() == R.id.account
                    || navDestination.getId() == R.id.profile_edit
                    || navDestination.getId() == R.id.phone_auth_check) {
                getWindow().setStatusBarColor(whiteColor);
                getWindow().setNavigationBarColor(whiteColor);
                uiVisibility |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
                        | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
            }
            // 다른 화면
            else {
                if (navDestination.getId() == R.id.movie_list) {
                    getWindow().setNavigationBarColor(ContextCompat.getColor(getApplication(), R.color.colorPrimary));
                } else {
                    getWindow().setNavigationBarColor(whiteColor);
                }
                getWindow().setStatusBarColor(blue700Color);
                uiVisibility |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
            }
            getWindow().getDecorView().setSystemUiVisibility(uiVisibility);
        });
    }

    public void showSnackbar(@NonNull String message) {
        Snackbar snackbar = Snackbar.make(binding.getRoot(), message, Snackbar.LENGTH_LONG);
        // 시작위치 Top 으로 변경
        View view = snackbar.getView();
        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) view.getLayoutParams();
        params.gravity = Gravity.TOP;
        view.setLayoutParams(params);

        // custom layout 으로 변경
        Snackbar.SnackbarLayout layout = (Snackbar.SnackbarLayout) snackbar.getView();
        View snackView = LayoutInflater.from(this).inflate(R.layout.snackbar, null);
        TextView tvMessage = snackView.findViewById(R.id.tv_message);
        tvMessage.setText(message);
        layout.addView(snackView);

        snackbar.setBackgroundTint(ContextCompat.getColor(getApplication(), R.color.WHITE));
        snackbar.show();
    }
}

File: app/src/main/java/com/hwonchul/movie/presentation/view/MainActivity.kt
package com.hwonchul.movie.presentation.view

import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.WindowInsetsController
import android.widget.FrameLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.navigation.fragment.NavHostFragment
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.Snackbar.SnackbarLayout
import com.hwonchul.movie.R
import com.hwonchul.movie.databinding.ActivityMainBinding
import com.hwonchul.movie.util.NetworkStatusHelper
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding

    @JvmField
    @Inject
    var helper: NetworkStatusHelper? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)

        setContentView(binding.root)
        setNetworkStatusHelper()
        setStatusBarAndNavigationBar()
    }

    private fun setNetworkStatusHelper() {
        helper?.observe(this) { networkStatus ->
            // 인터넷 연결됨
            if (networkStatus == NetworkStatusHelper.NetworkStatus.AVAILABLE) {
                val greenColor =
                    ContextCompat.getColor(application, android.R.color.holo_green_dark)
                binding.tvNetworkState.setBackgroundColor(greenColor)
                binding.tvNetworkState.setText(R.string.all_network_connected)
                Handler(Looper.getMainLooper()).postDelayed({
                    binding.tvNetworkState.visibility = View.GONE
                }, 2000)
            } else if (networkStatus == NetworkStatusHelper.NetworkStatus.UNAVAILABLE) {
                val redColor = ContextCompat.getColor(application, android.R.color.holo_red_dark)
                binding.tvNetworkState.setBackgroundColor(redColor)
                binding.tvNetworkState.setText(R.string.all_network_disconnected)
                binding.tvNetworkState.visibility = View.VISIBLE
            }
        }
    }

    private fun setStatusBarAndNavigationBar() {
        val navHostFragment =
            supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment
        val navController = navHostFragment.navController

        navController.addOnDestinationChangedListener { _, navDestination, _ ->
            val whiteColor = ContextCompat.getColor(application, R.color.WHITE)
            val blackColor = ContextCompat.getColor(application, R.color.BLACK)
            val blueColor = ContextCompat.getColor(application, R.color.blue)
            val blue700Color = ContextCompat.getColor(application, R.color.colorPrimary)

            when (navDestination.id) {
                R.id.poster -> updateSystemBarColor(blackColor, blackColor)
                R.id.login, R.id.phone_auth, R.id.account, R.id.profile_edit, R.id.phone_auth_check, R.id.nick ->
                    updateSystemBarColor(
                        whiteColor,
                        whiteColor,
                        isLightStatusBar = true,
                        isLightNavigationBar = true
                    )

                R.id.splash -> updateSystemBarColor(blueColor, whiteColor)
                R.id.movie_list -> updateSystemBarColor(blue700Color, blue700Color)
                R.id.movie_detail -> updateSystemBarColor(
                    blue700Color,
                    whiteColor,
                    isLightNavigationBar = true
                )
            }
        }
    }

    private fun updateSystemBarColor(
        statusBarColor: Int,
        navigationBarColor: Int,
        isLightStatusBar: Boolean = false,
        isLightNavigationBar: Boolean = false,
    ) {
        window.statusBarColor = statusBarColor
        window.navigationBarColor = navigationBarColor

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.insetsController?.apply {
                if (isLightStatusBar) {
                    setSystemBarsAppearance(
                        WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS,
                        WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS
                    )
                } else {
                    setSystemBarsAppearance(
                        0,
                        WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS
                    )
                }

                if (isLightNavigationBar) {
                    setSystemBarsAppearance(
                        WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS,
                        WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS
                    )
                } else {
                    setSystemBarsAppearance(
                        0,
                        WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS
                    )
                }
            }
        } else @Suppress("DEPRECATION") {
            var uiVisibility = window.decorView.systemUiVisibility
            uiVisibility = if (isLightStatusBar) {
                uiVisibility or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
            } else {
                uiVisibility and View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR.inv()
            }
            uiVisibility = if (isLightNavigationBar) {
                uiVisibility or View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
            } else {
                uiVisibility and View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR.inv()
            }
            window.decorView.systemUiVisibility = uiVisibility
        }
    }


    fun showSnackbar(message: String) {
        val snackbar = Snackbar.make(binding.root, message, Snackbar.LENGTH_LONG)
        // 시작위치 Top 으로 변경
        val view = snackbar.view
        val params = view.layoutParams as FrameLayout.LayoutParams
        params.gravity = Gravity.TOP
        view.layoutParams = params

        // custom layout 으로 변경
        val layout = snackbar.view as SnackbarLayout
        val snackView = LayoutInflater.from(this).inflate(R.layout.snackbar, null)
        val tvMessage = snackView.findViewById<TextView>(R.id.tv_message)
        tvMessage.text = message
        layout.addView(snackView)
        snackbar.setBackgroundTint(ContextCompat.getColor(application, R.color.WHITE))
        snackbar.show()
    }
}